src/decoder/lattice-incremental-decoder.h

```diff
@@ -0,0 +1,652 @@
+  // of it as a `pre-final state`
+  // Similarly, we define a `initial arc` as an arc from a initial-state, and the
+  // destination state of it as a `post-initial state`
+  // The post-initial states are constructed corresponding to pre-final states
+  // in the determinized and appended lattice before this chunk
+  // The pre-final states are constructed correponding to tokens in the last frames
+  // of this chunk.
+  // Since the StateId can change during determinization, we need to give permanent
+  // unique labels (as olabel) to these
+  // raw-lattice states for latter appending.
+  // We give each token an olabel id, called `token_label`, and each determinized and
+  // appended state an olabel id, called `state_label`
+  // step 2: Determinize the chunk of above raw lattice using determinization
+  // algorithm the same as LatticeFasterDecoder. Benefit from above `state_label` and
+  // `token_label` in initial and final arcs, each pre-final state in the last chunk
+  // w.r.t the initial arc of this chunk can be treated uniquely and each token in
+  // the last frame of this chunk can also be treated uniquely. We call the
+  // determinized new
+  // chunk `compact lattice (clat)`
+  // step 3: Appending the new chunk `clat` to the determinized lattice
+  // before this chunk. First, for each StateId in clat except its
+  // initial state, allocate a new StateId in the appended
+  // compact lattice. Copy the arcs except whose incoming state is initial
+  // state. Secondly, for each initial arcs, change its source state to the state
+  // corresponding to its `state_label`, which is a determinized and appended state
+  // Finally, we make the previous final arcs point to a "dead state"
+  // step 4 (optional): We re-determinize the appended lattice if needed.
+  //
+  // In our implementation, step 1 is done in GetIncrementalRawLattice(),
+  // step 2-4 is taken care by the class
+  // LatticeIncrementalDeterminizer
+  //
+  //   @param [in]  use_final_probs  If true *and* at least one final-state in HCLG
+  //                         was active on the final frame, include final-probs from
+  //                         HCLG
+  //                         in the lattice.  Otherwise treat all final-costs of
+  //                         states active
+  //                         on the most recent frame as zero (i.e.  Weight::One()).
+  //   @param [in]  redeterminize    If true, re-determinize the CompactLattice
+  //                         after appending the most recently decoded chunk to it,
+  //                         to
+  //                         ensure that the output is fully deterministic.
+  //                         This does extra work, but not nearly as much as
+  //                         determinizing
+  //                          a RawLattice from scratch.
+  //   @param [in]  last_frame_of_chunk  Pass the last frame of this chunk to
```

Comment:
Hm, so it looks like the prune-interval and determinize-delay are actually used outside this class?
If so, the config class should clearly document this, e.g. in a comment.
Make sure this works correctly if it's called twice on the same frame, we want it robust to user behavior.